Khatib amjad/issue25

khatibAmjad opened the pull request:
- **Created an endpoint for getting user's products with filter and search features**

1. added a router in the `src/routes/user.js` file.
2. added a controller called `getUserProducts` in the `src/controllers/user.js`
3. added swagger documentation in the `src/docs/User/getUserProducts.yaml`



- **Added user Id to the JWT token of the 3rd party authentication in `src/controllers/auth.js` file in the `saveUserToTokenAndCookie` controller, so the user can be fetched form the database in the get profile endpoint**


- **Reorganized yaml files inside the `src/docs` directory so the documentation is distributed like the controllers and the routes**


- **Changed the tags inside the yaml files so the swagger documentations are sorted in a proper way**


- **Unified the catch blocks inside all the controllers**


- **Created a nodemon.json file that'll allow the swagger doc page to refresh without needing to close and reopen the server**

khatibAmjad committed:
create getUserPosts router and controller

khatibAmjad committed:
fix issue25

khatibAmjad committed:
fix issue25

okurezgi commented on src/docs/Auth/register.yaml:
```diff
@@ -127,6 +125,8 @@ components:
           type: string
         refugee:
           type: boolean
+        avaregeRating:
```
we discharged these parts in the new schema design, can you delete these parts again?

khatibAmjad replied:
Woops sorry i's a merge mistake

okurezgi commented on src/docs/Auth/register.yaml:
```diff
@@ -143,6 +143,10 @@ components:
           type: string
         address:
           type: object
+        reviews:
```
same question here for reviews and reports

okurezgi commented on src/docs/User/getUserProducts.yaml:
```diff
@@ -0,0 +1,33 @@
+                  schema:
+                    type: array
+                    items:
+                      type: string
```
this part is from an enum so i think you can add an example here like ["Donated", "Requested", "Received"] to make it clear while using swagger

khatibAmjad replied:
Sure thing I'll add an enum to the doc

okurezgi commented on src/controllers/user.js:
```diff
@@ -12,7 +14,56 @@ const getProfile = async (req, res) => {
+  }
+};
+
+const getUserProducts = async (req, res) => {
```
In my opinion, this code is too complicated. If we discharge the part that user can search according to title, we can use UserModel to find received, requested and donated - it will be much cleaner. 
But if we keep it like this, I think we can also add filtering according to category.

khatibAmjad replied:
I think you're right about the code not being clean and as you said if we get rid of the search it'll be cleaner, but issue entailed making search feature. The thing here is the code that bad that we remove a feature or should we keep it.
for the category I don't agree that we should add it, let's discuss it on slack imo.

okurezgi commented on src/controllers/user.js:
```diff
@@ -12,7 +14,56 @@ const getProfile = async (req, res) => {
+      if (typeof postType !== 'object') postType = postType.split();
+
+      postType.forEach((type) => {
+        if (type === 'donated') {
```
we can uppercase the first letters for donated, received and requested

khatibAmjad replied:
Sure you're right

okurezgi reviewed (changes requested):
Great job kankim. I had few comments about naming and suggestions about the code, we can discuss about these parts.

khatibAmjad committed:
change naming in the controller. add publishere to products id. delete old schema fields from register swagger. change porp to PORT in app.js

khatibAmjad committed:
change variables naming in getUserProducts controllers. Add enum to the documentation

khatibAmjad committed:
Change variables' names in getUserProducts controller. Add description to swagger paremeters

khatibAmjad committed:
Change paths to regex expressions in constants file

okurezgi commented on src/controllers/user.js:
```diff
@@ -12,7 +15,62 @@ const getProfile = async (req, res) => {
+        publisher: req.params.userId,
+      });
+
+      return res.json(userPosts);
```
kankam shouldn't you add status code here also?

khatibAmjad replied:
I'll write them and push again

okurezgi commented on src/controllers/user.js:
```diff
@@ -12,7 +15,62 @@ const getProfile = async (req, res) => {
+    }
+
+    const userPosts = await ProductModel.find(finalQuery);
+    return res.json(userPosts);
```
also here?

okurezgi reviewed (approved):
Kankam good job. I just added two comment.

khatibAmjad committed:
Add status code to responses in getUserProducts controller

halitbatur commented on src/controllers/user.js:
```diff
@@ -12,7 +15,62 @@ const getProfile = async (req, res) => {
+    const finalQuery = { $and: [] };
+
+    if (postType) {
+      if (typeof postType !== 'object') postType = postType.split();
```
what if postType is an Object? in that case the .forEach does not work correctly as expected

halitbatur commented on src/controllers/user.js:
```diff
@@ -12,7 +15,62 @@ const getProfile = async (req, res) => {
+      if (typeof postType !== 'object') postType = postType.split();
+
+      postType.forEach((type) => {
+        if (type === 'Donated') {
```
instead of the reptetive if condition, objects can be used in order to change the value of postTypeDepending on the type so for example you can have an object like this 
```
POST_TYPE_SELECTOR = {
Donated: {postType:'Donate'},
Requested:{postType:'Request', isTransactionCompleted:false}
}
```
and then use this selector and spread the value of it inside the filteringQueries
like this:
```
      filteringQueries.$or.push({
            publisher: req.params.userId,
            {...POST_TYPE_SELECTOR[type]}
           });
```
now its a little bit clearner and whenever you have a new option u just add it to the SELECTOR object and its handeled automatically without needing to add another if condition

@okurezgi  @khatibAmjad  Hopefully this makes it a bit easier and cleaner

khatibAmjad committed:
Add POST_TYPE_SELECTOR to the constants and use it to replace the if statements in the controller. Add error responses to empty queries. Add new error to the swagger docs.

khatibAmjad committed:
Add an extra line after the last line in yaml file

khatibAmjad committed:
change homepage to swagger
merge with dev branch
fix login documentation
move logout documentation to a seperate file
move create product documentation to product folder

khatibAmjad committed:
change addNewProduct documentation file name
